scrapy/downloadermiddlewares/redirect.py

```diff
@@ -60,12 +63,14 @@ def _redirect(
         if ttl and redirects <= self.max_redirect_times:
             redirected.meta["redirect_times"] = redirects
             redirected.meta["redirect_ttl"] = ttl - 1
-            redirected.meta["redirect_urls"] = request.meta.get("redirect_urls", []) + [
-                request.url
-            ]
-            redirected.meta["redirect_reasons"] = request.meta.get(
-                "redirect_reasons", []
-            ) + [reason]
+            redirect_urls = request.meta.get("redirect_urls", []) + [request.url]
+            redirected.meta["redirect_urls"] = redirect_urls
+            redirect_reasons = request.meta.get("redirect_reasons", []) + [reason]
+            redirected.meta["redirect_reasons"] = redirect_reasons
+            fingerprints = request.meta.get("redirect_fingerprints", set())
+            assert self.crawler.request_fingerprinter is not None
+            fingerprint = self.crawler.request_fingerprinter.fingerprint(request)
+            redirected.meta["redirect_fingerprints"] = fingerprints | {fingerprint}
```

Comment:
The approach looks good to me.
However, I wonder if it'd cause issues in practice, in case someone uses `request.meta.copy()`. Unlike most other meta keys, it seems this one may affect the behavior significantly.

Do you think there is a room for making it less likely to happen? E.g. try to clear redirect_fingerprints once the response without a redirect is received? Or would it cause more edge cases than it'd solve?

Reply:
If we clean it in scenarios where the redirect middlewares return `response`, i.e. do not handle a redirect, there is the problem that the fact that one middleware does not handle a redirect does not mean that the other does not. So I discarded that approach.

I thought about removing them from the first of the middlewares (last on response processing), but then if that middleware is disabled, the cleanup would not happen.

We could also create a middleware specifically for the purpose of this cleanup. It feels weird, but it also feels like the cleanest approach to deal with this middleware duality and optionality.

Any thoughts or additional ideas?